### PR TITLE
feat: metrics aggregate

### PR DESCRIPTION
**Pull request by zreigz**
- Implement a controller to reconcile a metrics aggregate.
 - Create DebounceReconciler as a wrapper that fulfills the controller-runtime interface. It can be use for every controller to delay requests.

### CODE REVIEW

File: api/v1alpha1/metricsaggregate_types.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// +kubebuilder:subresource:status
+type MetricsAggregate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec of the MetricsAggregate
+	// +kubebuilder:validation:Required
+	Spec MetricsAggregateSpec `json:"spec"`
+
+	// Status of the IngressReplica
```
Comment (michaeljguarino) [inline api/v1alpha1/metricsaggregate_types.go:32]: comment for IngressReplica here

File: api/v1alpha1/metricsaggregate_types.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	// +kubebuilder:validation:Optional
+	Status Status `json:"status,omitempty"`
+}
+
+type MetricsAggregateSpec struct {
+	Nodes int `json:"nodes"`
```
Comment (michaeljguarino) [inline api/v1alpha1/metricsaggregate_types.go:38]: are these fields optional?  Feels like this should probably be the status struct tbh

File: internal/controller/metricsaggregate.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		metrics.Spec.MemoryAvailableBytes += nm.MemoryAvailableBytes
+		metrics.Spec.MemoryTotalBytes += nm.MemoryTotalBytes
+	}
+
+	fraction := float64(metrics.Spec.CPUTotalMillicores) / float64(metrics.Spec.CPUAvailableMillicores) * 100
+	metrics.Spec.CPUUsedPercentage = int64(fraction)
+	fraction = float64(metrics.Spec.MemoryTotalBytes) / float64(metrics.Spec.MemoryAvailableBytes) * 100
+	metrics.Spec.MemoryUsedPercentage = int64(fraction)
+
+	return ctrl.Result{}, reterr
```
Comment (michaeljguarino) [inline internal/controller/metricsaggregate.go:118]: shouldn't you be requeueing to update on an interval

**Review by michaeljguarino — COMMENTED** [review]
minor comments but makes sense overall

**Comment by michaeljguarino** [comment]
Another thing we should do is have the agent insert a default metrics aggregate resource on boot (if it doesn't exist).  I'm not sure we'll be able to do it in the helm chart smoothly, because of annoying CRD application/validation stuff, but it would be nice to have zero-config guarantees its there so the console server can query it.

Can just name the default one as `global` or something.

**Comment by zreigz** [comment]
> Another thing we should do is have the agent insert a default metrics aggregate resource on boot (if it doesn't exist). I'm not sure we'll be able to do it in the helm chart smoothly, because of annoying CRD application/validation stuff, but it would be nice to have zero-config guarantees its there so the console server can query it.
> 
> Can just name the default one as `global` or something.

The `DebounceReconciler` sends empty requests when there is no object. I have added `initGlobalMetricsAggregate` method in the `Reconcile` to discover it and create the global MetricsAggregate object

File: api/v1alpha1/metricsaggregate_types.go (hunk 1 above)
Comment (zreigz) [reply api/v1alpha1/metricsaggregate_types.go]: done

File: api/v1alpha1/metricsaggregate_types.go (hunk 2 above)
Comment (zreigz) [reply api/v1alpha1/metricsaggregate_types.go]: I've moved to status

File: internal/controller/metricsaggregate.go (hunk 3 above)
Comment (zreigz) [reply internal/controller/metricsaggregate.go]: done

**Review by michaeljguarino — COMMENTED** [review]
lgtm, why is the lint action failing though?

**Comment by zreigz** [comment]
Expects to bump the chart version

**Comment by zreigz** [comment]
> chart

it's a chart linter that expects to update the version, we do this only for releases

**Comment by zreigz** [comment]
@michaeljguarino can we merge this?

**Review by michaeljguarino — APPROVED** [review]

**Pull request merged**